### ISSUE TITLE
Illegal Memory access on GPU

### ISSUE THREAD

**Issue by aniquetahir** [deepmind/dm-haiku#204]
Hi, 
While training a simple MLP, I encountered an illegal memory access issue while running on GPU. The code runs fine on CPU. 

Relevant code:
```python
loss_fn_t = hk.transform(loss_fn)
loss_fn_t = hk.without_apply_rng(loss_fn_t)
...
loss_grad_fn = jax.grad(loss_fn_t.apply)
grads = loss_grad_fn(params, x, y)
```

Error Message:
```
JaxStackTraceBeforeTransformation: RuntimeError: Internal: Failed to load in-memory CUBIN: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered

The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------


The above exception was the direct cause of the following exception:

UnfilteredStackTrace                      Traceback (most recent call last)

UnfilteredStackTrace: RuntimeError: Internal: Failed to load in-memory CUBIN: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------


The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)

/usr/local/lib/python3.7/dist-packages/jax/interpreters/xla.py in _execute_compiled_primitive(prim, compiled, result_handler, *args)
    388   device, = compiled.local_devices()
    389   input_bufs = list(it.chain.from_iterable(device_put(x, device) for x in args if x is not token))
--> 390   out_bufs = compiled.execute(input_bufs)
    391   check_special(prim.name, out_bufs)
    392   return result_handler(*out_bufs)

RuntimeError: Internal: Failed to load in-memory CUBIN: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
```

Haiku version: dm_haiku-0.0.4-py3
Colab Notebook with reproduction of issue(**Please change runtime to GPU**):
https://colab.research.google.com/drive/1YN3kuM2skiMEbI-w3zX9Fqc-Jwon-kPT?usp=sharing

**Comment by tomhennigan** [comment]
Thank you for the clear reproducer!

I don't think this is Haiku specific. This seems to be an issue with either JAX or (I think more likely) the kernel we are using from NVIDIA. The following small reproducer (distilled from your code) has the same issue:

```python
import jax
import jax.numpy as jnp

a = jnp.ones([100, 100])
b = jnp.ones([100, 1])
c = jax.lax.dot_general(a, b, dimension_numbers=(((0,), (0,)), ((), ())))
c.block_until_ready()
```

It seems you can fix this by changing the second dim of `x` (in your program) to something other than `1`.

I've filed google/jax#7752 and hopefully someone can take a look next week.

**Comment by hawkinsp** [comment]
I looked into this on the JAX end. My conclusion is that this is an NVIDIA CUDA 11.0 bug in `ptxas`. It seems to be fixed in newer CUDA releases. The fix is to upgrade to a newer CUDA release. I verified that CUDA 11.2 and CUDA 11.4 both work. If upgrading is difficult, it's probably enough to grab the `ptxas` binary from a newer CUDA release and put it in your `PATH`.

Of course, it's a little hard to do that on Colab, but hopefully Colab upgrades sometime soon.
